New guide series for `arcgis.widgets` module

CMPeng opened the pull request:
All chapters required by https://github.com/ArcGIS/geosaurus/issues/6846 except for Part 6 are listed here. Part 6 (JupyterLab) does not render maps as expected and hence not to be come with this release.

Notebooks are placed inside `10-mapping-and-visualization-alt` folder. Please feel free to specify a better name if needed. @AtmaMani 

@BP-Ent Can you please do us a big favor by reviewing this guide series? Many thanks!!

BP-Ent commented on guide/10-mapping-and-visualization-alt/part7_utility_functions_for_the_map_widget.ipynb:
```diff
@@ -0,0 +1,357 @@
+{
```
<p>In this notebook, we will explore using utility functions to build interactivity. This can usually be split into two procedures: call back and display message. The <code>call back</code> function can be used to collect gis data on screen through digitization, while the <code>display message</code> function can be used to prompt users to perform a desired action. </p><p>To do this, we will need to setup an asynchronous <code>callback</code> function using the <code>on_click()</code> or <code>on_draw_end()</code> methods to create dynamic, interactive 'apps'. We will need to create a callback function, like <code>function_name(map_inst, geometry)</code>, with <code>map_inst</code> being the <code>MapView</code> instance and <code>geometry</code> being the <code>geometry</code> instance that the user is clicking. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

BP-Ent replied:
<h2>Demonstrate callback features with <code>on_click</code> and <code>on_draw_end</code> </h2><p>Our first example will take a point that a user clicks on the map, reverse geocode it from the geometry, and print out the resulting location. </p><p>Note, you can either create the GIS connection using an existing profile, or you can simply enter your username and password, e.g. gis = GIS("<a href="https://www.arcgis.com%22/" target="_blank">https://www.arcgis.com"</a>, "username", "password").</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

BP-Ent replied:
<p>Next, we want a user to draw a freehand <code>polyline</code> to indicate the paths they take for their runs. When the drawing operation ends, we use the <code>GIS</code>'s Geometry service to compute the length of the drawn path. We can do this by adding an event listener to the map widget that gets called when drawing is completed (i.e. <code>on_draw_end</code>). The event listener then computes the geodesic length of the drawn geometry using the geometry service and prints it out:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

BP-Ent replied:
<h2>Show message in the map widget via <code>display_message</code></h2><p> </p><p> </p><p><strong>Note:</strong> Only one message can be sent at a time, and multiple messages will not show up. Let's experiment with the first example by changing the <code>print</code> method to <code>display_message</code> so the desired text output can be shown inside the widget. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

BP-Ent replied:
<p>When <code>hide_mode_switch</code> is set to <code>True</code>, the 2D/3D switch button will be hidden from the widget.  </p><p> </p><p>  <strong>Note:</strong> Once the button is hidden, it cannot be made visible again. A new <code>MapView</code> instance must be created to see the button. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

AtmaMani replied:
<p>Since the JupyterLab part is not ready yet, let us rename part 7 to part 6, so it flows with the rest of the parts. Please rename title and filename accordingly</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

BP-Ent commented on guide/10-mapping-and-visualization-alt/part7_utility_functions_for_the_map_widget.ipynb:
```diff
@@ -0,0 +1,357 @@
+{
```
<p>The <code>set_js_cdn</code> function is called before the creation of any <code>MapView</code> object, and each instantiated object will use the specified <code>js_cdn</code> parameter as the ArcGIS API for JavaScript <code>CDN</code> URL, rather than the default <code>http://js.arcgis.com/4.X/</code>. This functionality is necessary in disconnected environments if the portal you are connecting to does not ship with the minimum necessary JavaScript API version. </p><p> </p><p><strong>Note:</strong> You may not need to call this function to view the widget in disconnected environments. For instance, if your computer cannot reach <code>js.arcgis.com</code> and you have a <code>GIS</code> connection to a portal, the widget will automatically attempt to use that Portal’s JS API that it ships with. </p><p> </p><p>For example, if <code>gis.map()</code> does not return a map, run the following code to configure the map widget with the Javascript API shipped with the portal:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

AtmaMani replied:
<p>"if the enterprise you are connecting " -- capitalize E in Enterprise (since its a product)</p>

BP-Ent commented on guide/10-mapping-and-visualization-alt/part7_utility_functions_for_the_map_widget.ipynb:
```diff
@@ -0,0 +1,357 @@
+{
```
<p>In Part 7 of this guide series, we discussed the utility functions of the MapView object, particularly in regard to using the callback and display_message mechanisms in creating interactivity for users. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

AtmaMani replied:
<p>Need to change Part 7 to 6 after renaming</p>

BP-Ent reviewed (changes requested):
Suggested changes made on reviewNB

CMPeng commented:
Thank you so much, @BP-Ent ! Your help is truly appreciated!!
All of your suggestions have been addressed. CC @AtmaMani

AtmaMani commented:
<p>Please implement Breck's suggestion here</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AtmaMani commented:
<p>Good  job!</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AtmaMani reviewed (changes requested):
@CMPeng thank you for this excellent series on the map widget. I made some suggestions via reviewnb app. Please take a look.
Also, thanks for addressing the issues in the networking guide.

AtmaMani commented:
For now, let us keep these files in the original folder numbered 10. The new site theme would allow us to group these separately from the rest of the mapping guides. We can do that in a future release

CMPeng commented:
<p>Hi @AtmaMani, this is a gif dynamically showing the map being dragged and panned, and looks OK on my local computer though. Let me upload it once again and see if works.</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

CMPeng commented:
<p>Atma, per our previous discussion, the side_by_side2 display will be a gif, and the other two will stay static. Do you want me to also change this one to a gif?</p>

---

 <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1142/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

CMPeng commented:
> @CMPeng thank you for this excellent series on the map widget. I made some suggestions via reviewnb app. Please take a look. Also, thanks for addressing the issues in the networking guide.

Thanks Atma!! I have addressed your comments and fixed the broken gif animations. Please check the latest commits!

CMPeng commented:
> @CMPeng one more (hopefully final) set of edits: Can you move the image, gif files (only the couple of files in this PR, not the ones already embedded) to https://github.com/Esri/arcgis-python-api/tree/master/static folder and update your links? Images need to be either base64 embedded or in the static folder for it to work reliably in old and new build systems.

Sure Atma! Please see the latest set of changed files. I have moved the gif and other images over to the static img folder, and corrected the reference accordingly. Thanks!
